### PR TITLE
brillo: init at 1.4.8

### PR DESCRIPTION
**Pull request by alexarice**
###### Motivation for this change
I wanted to use the brillo tool for backlight control on wayland but i could not find it.

###### Things done
Added brillo to nixpkgs/pkgs/os-specific/linux/brillo and added myself as a maintainer of it.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

I have not tried to add a nix package before so please let me know if I have messed something up.

### CODE REVIEW

**Comment by alexarice** [comment]
Have just realised this is not putting the udev rule in the correct location. Will try to fix soon

File: pkgs/os-specific/linux/brillo/default.nix
Comment (symphorien) [inline pkgs/os-specific/linux/brillo/default.nix]: Please use `fetchFromGitLab` instead. IIRC, the straightforward urls do not have a stable hash.

File: pkgs/os-specific/linux/brillo/default.nix
Comment (symphorien) [inline pkgs/os-specific/linux/brillo/default.nix]: Can you briefly explain the rationale for the patch either in the patch header or in a comment here?
Installing apparmor rules seems harmless to me; NixOS even has (some) apparmor support.

File: pkgs/os-specific/linux/brillo/default.nix
Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: Didn't know about this function, will change now

File: pkgs/os-specific/linux/brillo/default.nix
Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: I've just made another commit which changes all this

Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: should I add something along the lines of security.apparmor.packages = [pkgs.brillo] ?

File: pkgs/os-specific/linux/brillo/default.nix
Comment (jtojnar) [inline pkgs/os-specific/linux/brillo/default.nix]: Please do not use `DESTDIR`, it makes the path opaque to the project. `"PREFIX=${placeholder ''out''}"` should be enough. If something does not work such as AppArmor directory, you can override it individually.

File: pkgs/os-specific/linux/brillo/default.nix
Comment (jtojnar) [inline pkgs/os-specific/linux/brillo/default.nix]: Can you use patch instead? This will break silently when the rules file is changed.

File: pkgs/os-specific/linux/brillo/default.nix
Comment (jtojnar) [inline pkgs/os-specific/linux/brillo/default.nix]: The following is enough nowadays:

```suggestion
  pname = "brillo";
```

File: nixos/modules/hardware/brillo.nix
Comment (jtojnar) [inline nixos/modules/hardware/brillo.nix]: This might also need `environment.systemPackages` for polkit rules.

File: pkgs/os-specific/linux/brillo/default.nix
Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: I tried to stay very similar to the brightnessctl package, should this package be changed as well if I change this to use a patch instead of a command?

File: pkgs/os-specific/linux/brillo/default.nix
Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: are you suggesting I patch the makefile to change the AppArmor directory?

Comment (jtojnar) [reply pkgs/os-specific/linux/brillo/default.nix]: Changing it in `makeFlags` should work here.

Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: The AppArmor directory is set to `$(DESTDIR)/etc/apparmor.d` though so surely I can't change this without either patching or setting `DESTDIR` (I could just be misunderstanding makefiles though)

Comment (jtojnar) [reply pkgs/os-specific/linux/brillo/default.nix]: Variables set in `Makefile` will be still overwritten by those passed to `make`.

File: pkgs/os-specific/linux/brillo/default.nix
Comment (jtojnar) [reply pkgs/os-specific/linux/brillo/default.nix]: Yeah, we have a ton of `substitute`s that do not do anything in the tree and we should work on eliminating them. There is no hurry, though.

Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: How can you make a patch with the coreutils location in?

Comment (jtojnar) [reply pkgs/os-specific/linux/brillo/default.nix]: You create patch containing `@coreutils@` and use `substituteAll` to replace it.

Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: I've got something working but it doesn't look correct

Comment (jtojnar) [reply pkgs/os-specific/linux/brillo/default.nix]: Here is an example: https://github.com/NixOS/nixpkgs/blob/749c0a9c16e4ab6cee82884e949eb9fac93f6261/pkgs/applications/backup/deja-dup/default.nix#L20-L24

Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: ok I'll try that

File: nixos/modules/hardware/brillo.nix
Comment (alexarice) [reply nixos/modules/hardware/brillo.nix]: I have done this and now pkexec works but only if you input the full path (e.g. `pkexec /nix/store/ddavxh7mmbmxds6wpvqws1jw34j3nldh-brillo-1.4.3/bin/brillo -A 5`) I don't know if this is how it should be

Comment (alexarice) [reply nixos/modules/hardware/brillo.nix]: Should the polkit file be patched so that it lists the executable as something else

**Comment by dtzWill** [comment]
Please add an entry for this in `nixos/module-list.nix` otherwise the NixOS module isn't available for use! :)

**Comment by dtzWill** [comment]
And it looks like 1.4.4 is out, so might want to bump to that :).

**Comment by alexarice** [comment]
> And it looks like 1.4.4 is out, so might want to bump to that :).

I have bumped up the version number though for some reason I cannot work out it seems to be accepting the old sha256 hash. I don't think it should be doing this as if I purposefully break it then it says it is wrong and that it was expecting a hash which is not the one it is currently using. I've just checked and both hashes work in the derivation. Any idea what is causing this to happen and is it a problem @dtzWill

**Comment by alexarice** [comment]
I believe my above observation was due to caching on my laptop, it is now fixed.

**Comment by alexarice** [comment]
@dtzWill @jtojnar @symphorien is there anything else I need to do for this to be merged?

File: nixos/modules/hardware/brillo.nix
Comment (aanderse) [inline nixos/modules/hardware/brillo.nix]:
```
enable = mkEnableOption ''
  Enable brillo in userspace.
  This will allow brightness control from users in the video group.
'';
```

File: pkgs/os-specific/linux/brillo/default.nix
Comment (aanderse) [inline pkgs/os-specific/linux/brillo/default.nix]: Please be consistent in spacing: `{ stdenv, fetchFromGitLab , which, go-md2man, coreutils, substituteAll }:`

File: pkgs/os-specific/linux/brillo/default.nix
Comment (aanderse) [inline pkgs/os-specific/linux/brillo/default.nix]: To be consistent with spacing please change to: `nativeBuildInputs = [ go-md2man which ];`

**Review by aanderse — CHANGES_REQUESTED** [review]
I would ask that you please rewrite the git history as it has become rather difficult to follow. The first commit should be simply adding yourself to the maintainer list. After that you can probably lump all other changes into a single commit.

**Comment by alexarice** [comment]
I've made the changes you suggested and I think I've managed to rebase everything @aanderse though can't see how to mark as resolved

File: pkgs/os-specific/linux/brillo/default.nix
Comment (aanderse) [inline pkgs/os-specific/linux/brillo/default.nix]: Please correct the hash.

All my previous request for changes look resolved to me :+1:

**Review by aanderse — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/brillo/default.nix
Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: Sorry about that, must have crept in while rebasing

**Review by aanderse — APPROVED** [review]
:+1: LGTM
@jtojnar & @symphorien are you two satisfied to merge?

File: pkgs/os-specific/linux/brillo/default.nix
Comment (jtojnar) [inline pkgs/os-specific/linux/brillo/default.nix]: The indentation, order of attributes and their spacing is somewhat less idiomatic but the code itself looks fine to me.

**Comment by symphorien** [comment]
LGTM :)

File: pkgs/os-specific/linux/brillo/default.nix
Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: I'm happy to go through and fix this in a little bit, is there a style guide/example of a package that does it well?

Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: I've moved round some stuff and I think it's slightly nicer now

Comment (aanderse) [reply pkgs/os-specific/linux/brillo/default.nix]: @alexarice you accidentally added a third space right before `meta`. Other than that I think it looks great!

Comment (alexarice) [reply pkgs/os-specific/linux/brillo/default.nix]: fixed, hopefully I haven't managed to break something else this time

**Comment by aanderse** [comment]
@alexarice I'm testing this out on a laptop but can't seem to get anything to happen... 
`brillo -L` lists `intel_backlight`, but when I run any of the commands nothing seems to happen. Do you have a set of test commands you used that I can try?

**Comment by alexarice** [comment]
`brillo -A 0.5` and `brillo -U 0.5` works for me. Do you get an error message or does nothing happen? I have also just realised I've been running wayland and haven't tested it on X

**Comment by aanderse** [comment]
@alexarice I made a mistake and it turns out running `pkexec brillo -A 0.5` does work. If I run the command unprivileged (though my user has been added to the video group, logged out and back in), the command does not work and gives the following error:
```
brillo -A 0.5 -v 1
!ERR: src/file.c:191    file_read():    fopen: No such file or directory: '/home/andersea/.cache/brillo/backlight.intel_backlight.mincap'
!ERR: src/file.c:161    file_open():    open: Permission denied: '/sys/class/backlight/intel_backlight/brightness'
!ERR: src/main.c:24     main(): Execution failed
```

**Comment by alexarice** [comment]
Interesting, I've done a bit of investigation and I think the following is happening:
- I believe you are getting a permission denied as you don't have `services.udev.packages = [ brillo ];` in your `configuration.nix`. This should be set by the module but I don't know of a way to test modules without rebuilding my configuration off my local nixpkgs
- I looked at what was happening on mine and if I delete the file that doesn't exist on yours then It throws an error but works anyway
- I had set that file using `brillo -rc -S 1` but I had done this on 1.4.3 where it worked correctly
- I now have an error about failing to write to that file when running the same command
- If I manually do 'echo 1 >> backlight.intel_backlight.mincap` Then everything works correctly.
- The code throwing the error has changed from 1.4.3 to 1.4.4, Maybe this is an upstream bug?

**Comment by alexarice** [comment]
I have looked over the code and believe there could be an upstream error, should I revert this to 1.4.3 and submit an issue there?

**Comment by alexarice** [comment]
I have had a bit more of an investigate and have submitted a merge request upstream, though I believe at the moment the only functionality which doesn't work is the minimum brightness setting and this can be manually set

**Comment by aanderse** [comment]
>     * I believe you are getting a permission denied as you don't have `services.udev.packages = [ brillo ];` in your `configuration.nix`. This should be set by the module but I don't know of a way to test modules without rebuilding my configuration off my local nixpkgs

I installed the brillo module that you have here. I also tried bumping the version down to 1.4.3 and same error.
```
brillo -A 0.5 -v 1
!ERR: src/file.c:192    file_read():    fopen: No such file or directory: '/home/andersea/.cache/brillo/backlight.intel_backlight.mincap'
!ERR: src/file.c:163    file_open():    open: Permission denied: '/sys/class/backlight/intel_backlight/brightness'
!ERR: src/main.c:27     main(): Execution failed
```

I feel like I'm doing something wrong?

**Comment by dtzWill** [comment]
Just for a new data point-- we also have `light` (which brillo forked)
which you may want to try.

**Comment by alexarice** [comment]
That is very strange, I just built my system against this fork with `nixos-rebuild -I ~/nixpkgs switch` and added only `hardware.brillo.enable = true` and I get the following output:
```
brillo -A 5 -v 1 
!ERR: src/file.c:191	file_read():	fopen: No such file or directory: '/home/alex/.cache/brillo/backlight.intel_backlight.mincap'
```
But my screen gets brighter.

For the error above I am trying to get a fix upstream https://gitlab.com/cameronnemo/brillo/merge_requests/4

**Comment by alexarice** [comment]
Upstream issues have been fixed and I have moved this to 1.4.6.
@aanderse did you every figure out what was giving you permission errors?

**Comment by aanderse** [comment]
It looks like you might want to nudge upstream on that 1.4.6 release...
```
~> brillo -V
brillo 1.4.5
```

https://gitlab.com/cameronnemo/brillo/blob/v1.4.6/src/light.h#L8

The issue persists. I have to assume I'm doing something wrong at this point. As a KDE user I don't actually need this program... so I'm going to leave it up to someone else to say whether the program works or not. It would be nice if at least one person other than @alexarice could confirm this works under X11 as expected.

**Comment by alexarice** [comment]
Did you try if the program light works for you as that has near identical module definition. Have asked for the version to be bumped upstream

**Comment by aanderse** [comment]
> Did you try if the program light works for you as that has near identical module definition. Have asked for the version to be bumped upstream

@alexarice I have not and as I mentioned at this point I'm happy if someone else can confirm.

**Pull request merged**
